pulsar/source.go

```diff
@@ -0,0 +1,161 @@
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetValue(msg interface{}) []byte {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetOffset(msg interface{}) int64 {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) RegisterHook(hook SourceHook) {
+	p.hook = hook
+}
+
+func GetPulsarSource(conf PulsarConf) *PulsarSource {
+	return &PulsarSource{conf: conf}
+}
+
+// Generate is Source method implementation, which connects to Pulsar and pushes
+// PulsarMessage into the channel
+func (p *PulsarSource) Generate(out chan<- interface{}) {
+	// Prepare KeyFile
+	props := map[string]string{
+		"type":          "client_credentials",
+		"client_id":     p.conf.AuthClientId,
+		"client_secret": p.conf.AuthClientSecret,
+		"issuer_url":    p.conf.AuthIssuerURL,
+	}
+	privateKey, _ := json.Marshal(props)
+	fmt.Println("prepared pulsar privateKey", privateKey)
+
+	// Build authentication
+	auth := pulsar.NewAuthenticationOAuth2(map[string]string{
+		"type":       "client_credentials",
+		"issuerUrl":  p.conf.AuthIssuerURL,
+		"audience":   p.conf.AuthAudience,
+		"privateKey": fmt.Sprintf("data://%s", string(privateKey)),
+		"clientId":   p.conf.AuthClientId,
+	})
+	fmt.Println("prepared pulsar oauth2")
+
+	// Authenticate client
+	client, err := pulsar.NewClient(pulsar.ClientOptions{
+		URL:            p.conf.Url,
+		Authentication: auth,
+	})
+	fmt.Println("prepared client with authentication")
+	if err != nil {
+		panic(err)
+	}
+	var subsriptionType pulsar.SubscriptionType
+	if strings.EqualFold(p.conf.SubscriptionType, "KeyShared") {
+		subsriptionType = pulsar.KeyShared
+		log.Printf("starting with subscription type keyshared")
+	} else {
+		log.Printf("starting with subscription type failover")
+		subsriptionType = pulsar.Failover
+	}
+
+	// Open channel for consumer
+	channel := make(chan pulsar.ConsumerMessage, 100)
+	options := pulsar.ConsumerOptions{
+		Topic:            p.conf.Topic,
+		SubscriptionName: p.conf.SubscriptionName,
+		Type:             subsriptionType,
+	}
+	options.MessageChannel = channel
+	consumer, err := client.Subscribe(options)
+	if err != nil {
+		client.Close()
+		panic(err)
+	}
+
+	if p.conf.ForceRestart && p.conf.ReadNewest {
+
+		log.Println("Setting force restart as true and readnewest as true, the consumers will start listenning from time  = %v", time.Now().UTC())
+		//er := consumer.Seek(pulsar.EarliestMessageID())
+		er := consumer.SeekByTime(time.Now().UTC())
+		if er != nil {
+			client.Close()
+			panic(er)
+		}
+		log.Println("Done setting force restart as true and readnewest as true")
+	} else if p.conf.ForceRestart && p.conf.SeekByTime != 0 {
+		seekByTime := time.Duration(p.conf.SeekByTime) * time.Second
+		updatedTime := time.Now().UTC().Add(-seekByTime)
+		er := consumer.SeekByTime(updatedTime)
+		if er != nil {
+			client.Close()
+			panic(er)
+		}
+		fmt.Printf("Done Setting SeekByTime as %v %v\n", seekByTime.String(), updatedTime)
```

Comment:
Change to log.Println. Don't mix fmt and log.

Reply:
ack